src/System.Private.CoreLib/src/System/Globalization/FormatProvider.FormatAndParse.Unix.cs

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Runtime;
+using System.Globalization;
+using System.Diagnostics;
+
+namespace System.Globalization
+{
+    internal partial class FormatProvider
+    {
+        private partial class Number
+        {
+            // buffer size to hold digits (ECVT_MAX_COUNT_SIZE), decimal point, number sign, exponent, exponent symbol 'e', exponent sign and null.
+            private const int  ECVT_MAX_BUFFER_SIZE = 357;  
+            private const int  ECVT_MAX_COUNT_SIZE  = 348; // max digits count for floating numbers
+
+            internal static unsafe void _ecvt_s(byte* buffer, int sizeInBytes, double value, int count, int* dec, int* sign)
+            {
+                *dec = *sign = 0;
+                
+                if (buffer == null || sizeInBytes <= 0 || count < 0)
+                {
+                    return;
+                }
+                
+                buffer[0] = 0;
+
+                if (count > ECVT_MAX_COUNT_SIZE)
+                {
+                    count = ECVT_MAX_COUNT_SIZE;
+                }
+                
+                if (count > sizeInBytes - 1) // make sure we have room for null too
+                {
+                    count = sizeInBytes - 1;
+                }
+                
+                if (value < 0.0)
+                {
+                    *sign = 1;
+                }
+
+                if (value == 0.0)
+                {
+                    for (int j=0; j<count; j++)
+                    {
+                        buffer[j] = (byte) '0';
+                    }
+                    buffer[count] = 0;
+                    return;
+                }
+                
+                byte *tempBuffer = stackalloc byte[ECVT_MAX_BUFFER_SIZE];              
+                
+                //
+                // Get the number formatted as a string in the form x.xxxxxxexxxx
+                //
+
+                // "%.348e"
```

Comment:
If I understand this properly, this formatting string will produce 300+ hundred digits long string, but we are typically only interested in the first at most 20 characters. Is there no good way to make this more efficient?